### PR TITLE
Implement EditableTimeEntry

### LINKED ISSUE #55: Implement EditableTimeEntry
## Description
`EditableTimeEntry` functionality should be modelled after this PR: https://github.com/toggl/android/pull/79
## Definition of done
- [ ] The `EditableTimeEntry` type is added and is used in the `StartEditState` and `TimerState`

### PR DESCRIPTION
**Pull request by arturs105**
Closes #55 

The implementation is copied from toggl/android#79

### CODE REVIEW

File: Modules/Models/Models/Workspace.swift [hunk 1]
```diff
@@ -5,10 +5,25 @@ public struct Workspace: Codable, Entity, Equatable {
...
+    public init(id: Int, name: String, admin: Bool, features: [WorkspaceFeature]) {
+        self.id = id
+        self.name = name
+        self.admin = admin
+        self.features = features
+    }
+}
+
+extension Workspace {
+    var irPro: Bool {
```
Comment (heytherewill) [inline Modules/Models/Models/Workspace.swift]:
```suggestion
    var isPro: Bool {
```

File: Modules/Models/Models/WorkspaceFeature.swift [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import Foundation
+
+public enum WorkspaceFeature : Int, Codable {
```
Comment (Odrakir) [inline Modules/Models/Models/WorkspaceFeature.swift]: I don't know where does this come from and why do we need it. I think this shouldn't be part of this PR, right?

File: Modules/Models/Models/WorkspaceFeature.swift [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    case Free = 0
+    case Pro = 13
+    @available(*, deprecated, message: "Use specific granular features instead")
```
Comment (Odrakir) [inline Modules/Models/Models/WorkspaceFeature.swift]: Why are we adding code that's already deprecated?!

File: Modules/Repository/Repository/Repository.swift [hunk 4]
```diff
@@ -10,7 +11,7 @@ public protocol TimeLogRepository {
     func getProjects() -> Single<[Project]>
     func getTasks() -> Single<[Task]>
     func getTags() -> Single<[Tag]>
-    func startTimeEntry(timeEntry: TimeEntry) -> Single<(started: TimeEntry, stopped: TimeEntry?)>
+    func startTimeEntry(workspaceId: Int, description: String) -> Single<(started: TimeEntry, stopped: TimeEntry?)>
```
Comment (Odrakir) [inline Modules/Repository/Repository/Repository.swift]: Pretty soon will be adding other parameters here, doesn't it make sense to send a struct instead? If you don't want to use `TimeEntry` because it has an id, maybe we can create a `NewTimeEntry` struct or use the `EditableTimeEntry` moving the ids property out of it.

File: Modules/Repository/Repository/Repository.swift [hunk 5]
```diff
@@ -87,20 +90,29 @@ extension Repository: TimeLogRepository {
...
+    public func startTimeEntry(workspaceId: Int, description: String) -> Single<(started: TimeEntry, stopped: TimeEntry?)> {
         
-        var stoppedTimeEntry: TimeEntry?
-        if let runningEntryIndex = timeEntries.firstIndex(where: { $0.duration == 0}) {
-            stoppedTimeEntry = timeEntries[runningEntryIndex]
-            stoppedTimeEntry!.duration = timeEntry.start.timeIntervalSince(stoppedTimeEntry!.start)
-            timeEntries[runningEntryIndex] = stoppedTimeEntry!
+        let stoppedTimeEntry = stopRunningTimeEntry()
+        let startedTimeEntry = TimeEntry(
+            id: (timeEntries.map({ $0.id }).max() ?? 0) + 1,
```
Comment (Odrakir) [inline Modules/Repository/Repository/Repository.swift]: Des is also touching this code. This should be a `UUID`

File: Modules/Repository/Repository/Repository.swift [hunk 6]
```diff
@@ -87,20 +90,29 @@ extension Repository: TimeLogRepository {
...
+            description: description,
+            start: time.now(),
+            duration: -1,
```
Comment (Odrakir) [inline Modules/Repository/Repository/Repository.swift]: Not sure if this is supposed to be negative, 0 or `nil`... can you verify this?

File: Modules/Timer/Timer/Common/EditableTimeEntry.swift [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+import Models
+
+public struct EditableTimeEntry : Equatable {
+    public var ids: [Int]
+    public var workspaceId: Int
+    public var description: String
+    public var billable: Bool
+    
+    private init(
+        ids: [Int],
```
Comment (Odrakir) [inline Modules/Timer/Timer/Common/EditableTimeEntry.swift]: As I mentioned above, maybe having a struct to gather all edited data (`TimeEntryDTO` for example) and move the ids array out of it could allow us to use this struct when sending data to the repository in `startTimeEntry`

File: Modules/Timer/Timer/StartEdit/StartEditReducer.swift [hunk 8]
```diff
@@ -7,29 +7,22 @@ import OtherServices
 
 func createStartEditReducer(repository: TimeLogRepository, time: Time) -> Reducer<StartEditState, StartEditAction> {
     return Reducer {state, action in
-
+
```
Comment (Odrakir) [inline Modules/Timer/Timer/StartEdit/StartEditReducer.swift]: We will probably need another action to set the initial data when editing a TE. Not sure if it belongs to this PR, just mentioning it here.

File: Modules/Models/Models/Workspace.swift (hunk 1 above)
Comment (arturs105) [reply Modules/Models/Models/Workspace.swift]: This seems like a brain fart because `is` in Latvian is `ir` 😄

File: Modules/Models/Models/WorkspaceFeature.swift (hunk 3 above)
Comment (arturs105) [reply Modules/Models/Models/WorkspaceFeature.swift]: It was added in the droid repo as well. @heytherewill might know the reason

Comment (heytherewill) [reply Modules/Models/Models/WorkspaceFeature.swift]: Copied from the Xamarin repo. We might still receive this from API calls, but if we ever have to handle these in code, we should get notified and then discuss with backend how to proceed.

File: Modules/Repository/Repository/Repository.swift (hunk 4 above)
Comment (heytherewill) [reply Modules/Repository/Repository/Repository.swift]: for `startTimeEntry` it'd be good to have a `NewTimeEntry` struct that doesn't have a duration property, as described in this issue: https://github.com/toggl/mobileapp/issues/6452

File: Modules/Repository/Repository/Repository.swift (hunk 6 above)
Comment (arturs105) [reply Modules/Repository/Repository/Repository.swift]: If you check out the `master` branch, open up `StartEditReducer.swift` and go to `case .startTapped`, you can see the duration being set to -1. That's the reason i went with the same approach.

However, i think, it would be better if duration was nullable and was set to null for running time entries.

File: Modules/Timer/Timer/StartEdit/StartEditReducer.swift (hunk 8 above)
Comment (arturs105) [reply Modules/Timer/Timer/StartEdit/StartEditReducer.swift]: Good point, but definitely shouldn't be a part of this PR.

File: Modules/Models/Models/WorkspaceFeature.swift (hunk 3 above)
Comment (Odrakir) [reply Modules/Models/Models/WorkspaceFeature.swift]: Ok to this specific case. But please let's don't copy the other's platform code without questioning it, this is the perfect recipe for bugs.

File: Modules/Repository/Repository/Repository.swift (hunk 6 above)
Comment (Odrakir) [reply Modules/Repository/Repository/Repository.swift]: Ok, I'll add an issue

File: Modules/Models/Models/WorkspaceFeature.swift (hunk 3 above)
Comment (heytherewill) [reply Modules/Models/Models/WorkspaceFeature.swift]: Don't worry, I'm not doing it :)

File: Modules/Repository/Repository/TimeEntryDto.swift [hunk 9]
```diff
@@ -0,0 +1,18 @@
+import Foundation
+import Models
+
+public struct TimeEntryDto {
```
Comment (heytherewill) [inline Modules/Repository/Repository/TimeEntryDto.swift]: This name is not specific enough. Call it `StartTimeEntryDTO` or `NewRunningTimeEntryDTO`, but make sure that the name makes it explicit about its usage to prevent people from abusing it.

File: Modules/Models/Models/Workspace.swift [hunk 10]
```diff
@@ -5,25 +5,16 @@ public struct Workspace: Codable, Entity, Equatable {
...
-    public init(id: Int, name: String, admin: Bool, features: [WorkspaceFeature]) {
+    public init(id: Int, name: String, admin: Bool) {
         self.id = id
         self.name = name
         self.admin = admin
-        self.features = features
-    }
-}
-
-extension Workspace {
```
Comment (heytherewill) [inline Modules/Models/Models/Workspace.swift]: Why was this deleted? Are you adding it only in the PR that checks it?

Comment (heytherewill) [reply Modules/Models/Models/Workspace.swift]: I'm not arguing for it to be brought back, I'm merely questioning

Comment (arturs105) [reply Modules/Models/Models/Workspace.swift]: Yeah, it was removed because it's not related to this PR. Gonna add it when implementing the `toggleBillableTapped` action, since it's needed there.

File: Modules/TogglTrack/TogglTrack/App/AppState.swift [hunk 11]
```diff
@@ -12,7 +12,8 @@ public struct AppState {
     
     public var localLoadingState: LocalLoadingState = LocalLoadingState()
     public var localOnboardingState: LocalOnboardingState = LocalOnboardingState()
-    public var localTimerState: LocalTimerState = LocalTimerState()
+    public var localTimerState: LocalTimerState = LocalTimerState(
+        defaultWorkspace: Workspace(id: 1, name: "Auto created workspace", admin: true))
```
Comment (Odrakir) [inline Modules/TogglTrack/TogglTrack/App/AppState.swift]: This should never be done like this. I guess this  is temporal, but still I think it makes sense to move the default workspace to either the user property, its own property in the app state or as a selector from user and workspaces.

I think it's worth it to make that change in this PR because building the `LocalTimerState` like this is very wrong.

So, I would add one of these, which will be set on the `LoadingReducer`:
- `defaultWorkspace` property in `TimeLogEntities`
- `defaultWorkspace` property in `User`
- `defaultWorkspace` property in `AppState`

@heytherewill thoughts?

File: Modules/Timer/Timer/Common/EditableTimeEntry.swift [hunk 12]
```diff
@@ -0,0 +1,53 @@
+import Foundation
+import Models
+import Repository
+
+public struct EditableTimeEntry: Equatable {
```
Comment (Odrakir) [inline Modules/Timer/Timer/Common/EditableTimeEntry.swift]: Since we now have the `TimeEntryDTO` we don't need this struct anymore. We could have a `TimeEntryDTO` and an `[Int]` for the ids in the `StartEditState` instead. I think it's much cleaner. @heytherewill ?

Comment (arturs105) [reply Modules/Timer/Timer/Common/EditableTimeEntry.swift]: 🤔 That would work, yes. But my thinking was that the `StartTimeEntryDto` is immutable (notice the `let`s) and has a single purpose of passing data necessary to start a time entry. We'd later have other single purpose DTOs for other functionality. 

On the other hand, `EditableTimeEntry` is, well, editable, hence the `var` properties and is used for the single purpose of editing.

**tl;dr**: I think we should keep both because of single responsibility principle

File: Modules/TogglTrack/TogglTrack/App/AppState.swift (hunk 11 above)
Comment (arturs105) [reply Modules/TogglTrack/TogglTrack/App/AppState.swift]: Yeah, i copied this from the droid repo and had the same thought that this smells and is temporal so i should have questioned it, but i didn't in order to proceed at high velocity.

That being said, i support this change and we should pick one of the three options you're offering. Not sure which one though 🤔

File: Modules/Timer/Timer/Common/EditableTimeEntry.swift (hunk 12 above)
Comment (Odrakir) [reply Modules/Timer/Timer/Common/EditableTimeEntry.swift]: Immutability is not that important for a DTO, because once you pass it around every method has its own copy, it's ok if they mutate it, the caller won't see the changes. I still think reusing the DTO is cleaner, the single responsibility principle doesn't apply here since it's all the same responsibility of holding the TE's data, but I would like to hear @heytherewill opinion.

File: Modules/TogglTrack/TogglTrack/App/AppState.swift (hunk 11 above)
Comment (Odrakir) [reply Modules/TogglTrack/TogglTrack/App/AppState.swift]: Any would do, this is easy to change in the future. 🤔 . If you want me to decide... let's put it into the `TimeLogEntities` struct.

Comment (arturs105) [reply Modules/TogglTrack/TogglTrack/App/AppState.swift]: `TimeLogEntities ` struct it is then

File: Modules/Timer/Timer/Common/EditableTimeEntry.swift (hunk 12 above)
Comment (heytherewill) [reply Modules/Timer/Timer/Common/EditableTimeEntry.swift]: >  I still think reusing the DTO is cleaner
This caused a lot of problems in the old app, because then you can't ensure constraints in compile time. When you start a time entry, it can't have a duration and when supporting premium features there might be even more of those, so ensuring that during compile time is much better IMO. Also, it doesn't need an implementation of the protocol and doesn't give you false expectations about the usage of `startTimeEntry` (e.g.: Passing a DTO that has a duration and expecting it to create a time entry that's not running).

Bottomline: I'd use specific DTOs, like was agreed for the old apps but never implemented, rather than reusing an interface, which was a mistake made in the old app. And no, I'm not blindly saying we should avoid the fears of the old app; I really think this is the best way to design the api for the repo.

Comment (heytherewill) [reply Modules/Timer/Timer/Common/EditableTimeEntry.swift]: I just noticed that here you are commenting on the `EditableTimeEntry` and not on the start time entry DTO. I think `EditableTimeEntry` is still the way to go, though, because it's a concrete type with everything that we need rather than two properties that aren't immediately connected 🤔

Comment (Odrakir) [reply Modules/Timer/Timer/Common/EditableTimeEntry.swift]: I'm not sure I understand. I'm not saying `TimeEntryDTO` should be a protocol, it's a struct. Basically the same as `EditableTimeEntry`, but keeping the ids out of it in another property.

File: Modules/TogglTrack/TogglTrack/App/AppState.swift (hunk 11 above)
Comment (juanlaube) [reply Modules/TogglTrack/TogglTrack/App/AppState.swift]: I'd say it should be in `User` since it's a property of the user in the business logic. However in some particular cases the default workspace in the user's profile is `null` (probably an error in the API or something, I don't think that's intended) and we need to make a decision, in the old app we decided to pick the workspace with the lowest id if I remember correctly.

File: Modules/Timer/Timer/StartEdit/StartEditReducer.swift [hunk 13]
```diff
@@ -42,6 +37,9 @@ func createStartEditReducer(repository: TimeLogRepository, time: Time) -> Reduce
     }
 }
 
-func startTimeEntry(_ timeEntry: TimeEntry, repository: TimeLogRepository) -> Effect<StartEditAction> {
-    return Effect.empty
+func startTimeEntry(_ timeEntry: StartTimeEntryDto, repository: TimeLogRepository) -> Effect<StartEditAction> {
+    return repository
+        .startTimeEntry(timeEntry)
+        .toEffect(map: { (started, _) in StartEditAction.timeEntryAdded(started) },
```
[the remark this answers is not in the document]
Comment (arturs105) [reply Modules/Timer/Timer/StartEdit/StartEditReducer.swift:43]: No, not yet. That action is added in #33 and that PR depends on this one.

**Comment by arturs105** [comment]
Also: i'd rather add the `StartEditReducer ` tests in #33, since that PR changes the actions a bit.
